shared/storage/minio.py

```diff
@@ -1,29 +1,42 @@
 import gzip
 import json
 import logging
-import os
-import shutil
 import sys
-import tempfile
 from io import BytesIO
-from typing import BinaryIO, overload
+from typing import IO, BinaryIO, Tuple, cast, overload
 
+import zstandard
 from minio import Minio
-from minio.credentials import (
+from minio.credentials.providers import (
     ChainedProvider,
     EnvAWSProvider,
     EnvMinioProvider,
     IamAwsProvider,
 )
 from minio.deleteobjects import DeleteObject
 from minio.error import MinioException, S3Error
+from minio.helpers import ObjectWriteResult
+from urllib3 import HTTPResponse
 
-from shared.storage.base import CHUNK_SIZE, BaseStorageService
+from shared.storage.base import CHUNK_SIZE, PART_SIZE, BaseStorageService
 from shared.storage.exceptions import BucketAlreadyExistsError, FileNotInStorageError
 
 log = logging.getLogger(__name__)
 
 
+class GZipStreamReader:
+    def __init__(self, fileobj: IO[bytes]):
+        self.data = fileobj
+
+    def read(self, size: int = -1, /) -> bytes:
+        curr_data = self.data.read(size)
+
+        if not curr_data:
+            return b""
+
+        return gzip.compress(curr_data)
```

Comment:
this means you are compressing each chunk you read individually, which is not ideal from a compression point of view, but also means you end up with multiple concatenated gzip streams. I know zstd can handle that, though depending on the specific library you use, you have to opt-into this behavior. I’m not sure about gzip in this case.

I’m surprised you have to implement this yourself though. I would imagine gzip should have a streaming compressor built in?

Reply:
> this means you are compressing each chunk you read individually, which is not ideal from a compression point of view, but also means you end up with multiple concatenated gzip streams. I know zstd can handle that, though depending on the specific library you use, you have to opt-into this behavior. I’m not sure about gzip in this case.

this seems to be fine in the case of gzip: [link](https://www.gnu.org/software/gzip/manual/html_node/Advanced-usage.html)

> I’m surprised you have to implement this yourself though. I would imagine gzip should have a streaming compressor built in?

it doesn't follow the same interface as the python-zstandard streaming interface, and i couldn't come up with a way to code it that isn't awkward.